Error referencing a non-existent page destination when writing PDF

Errors occur in the PDF writing process.

## Environment

Which environment were you using when you encountered the problem?

```prompt
> python -m platform
Windows-10-10.0.22631-SP0

> python -c "import pypdf;print(pypdf._debug_versions)"
pypdf==4.3.1, crypt_provider=('cryptography', '41.0.5'), PIL=10.4.0
```

## Code + PDF

This is a minimal, complete example that shows the issue:

```python
from pypdf import PdfWriter, PdfReader

def test_write_pdf():
    filepath = r"C:\test.pdf"
    with open(filepath, "rb") as f:
        pdf_writer = PdfWriter()
        pdf_reader = PdfReader(f, True)
        print(pdf_reader.metadata)
        print(pdf_reader.named_destinations)
        pdf_writer.append(pdf_reader)
```

Sorry we are unable to provide the PDF.
We are in the process of confirming that we can create a PDF that can be published without any problems.

## Traceback

This is the complete traceback I see:

```
venv\venv\Lib\site-packages\pypdf\_writer.py:2365: in append
    self.merge(
venv\venv\Lib\site-packages\pypdf\_writer.py:2474: in merge
    p = reader.pages[dest["/Page"]]
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <pypdf._page._VirtualList object at 0x0000025E7C033650>, index = 1

    def __getitem__(
        self, index: Union[int, slice]
    ) -> Union[PageObject, Sequence[PageObject]]:
        if isinstance(index, slice):
            indices = range(*index.indices(len(self)))
            cls = type(self)
            return cls(indices.__len__, lambda idx: self[indices[idx]])
        if not isinstance(index, int):
            raise TypeError("sequence indices must be integers")
        len_self = len(self)
        if index < 0:
            # support negative indexes
            index = len_self + index
        if index < 0 or index >= len_self:
>           raise IndexError("sequence index out of range")
E           IndexError: sequence index out of range
```

Perhaps the following are causing the problem. `46 0 obj` and `20 0 obj` do not exist.
I tried to fix them in Adobe Acrobat, but could not figure out how to turn them off.

```
30 0 obj
<</AcroForm 46 0 R/Dests 20 0 R/Extensions<</ADBE<</BaseVersion/1.7/ExtensionLevel 8>>>>/Metadata 5 0 R/Names 47 0 R/OCProperties<</D<</OFF[]/Order[]/RBGroups[]>>/OCGs[48 0 R 49 0 R 50 0 R]>>/Pages 18 0 R/StructTreeRoot 14 0 R/Type/Catalog>>
endobj
```

When read in PdfReader, the following will be generated in `named_destinations`. (pdf has only one page)

```
{'/__WKANCHOR_2': {'/Title': '/__WKANCHOR_2', '/Page': 0, '/Type': '/XYZ', '/Left': 36, '/Top': 754, '/Zoom': 0.0}, '/__WKANCHOR_4': {'/Title': '/__WKANCHOR_4', '/Page': 0, '/Type': '/XYZ', '/Left': 305, '/Top': 754, '/Zoom': 0.0}, '/__WKANCHOR_6': {'/Title': '/__WKANCHOR_6', '/Page': 0, '/Type': '/XYZ', '/Left': 36, '/Top': 454, '/Zoom': 0.0}, '/__WKANCHOR_8': {'/Title': '/__WKANCHOR_8', '/Page': 1, '/Type': '/XYZ', '/Left': 61, '/Top': 802, '/Zoom': 0.0}, '/__WKANCHOR_a': {'/Title': '/__WKANCHOR_a', '/Page': 1, '/Type': '/XYZ', '/Left': 36, '/Top': 425, '/Zoom': 0.0}, '/__WKANCHOR_c': {'/Title': '/__WKANCHOR_c', '/Page': 2, '/Type': '/XYZ', '/Left': 36, '/Top': 814, '/Zoom': 0.0}, '/__WKANCHOR_e': {'/Title': '/__WKANCHOR_e', '/Page': 2, '/Type': '/XYZ', '/Left': 36, '/Top': 703, '/Zoom': 0.0}}
```
I was able to avoid the error by adding `if len(reader.pages) > dest[“/Page”]:` on the PdfWriter side.

https://github.com/py-pdf/pypdf/blob/8f621201210bad8fcd56bd50a037e1e9fe79c500/pypdf/_writer.py#L2471-L2482

```python
            elif isinstance(dest["/Page"], int):
                # the page reference is a page number normally not a PDF Reference
                # page numbers as int are normally accepted only in external goto
                if len(reader.pages) > dest["/Page"]:
                    p = reader.pages[dest["/Page"]]
                    assert p.indirect_reference is not None
                    try:
                        arr[NumberObject(0)] = NumberObject(
                            srcpages[p.indirect_reference.idnum].page_number
                        )
                        self.add_named_destination_array(dest["/Title"], arr)
                    except KeyError:
                        pass
```

A Dests reference is created in the resulting PDF as follows.

```
15 0 obj
<<
/Dests 16 0 R
>>
endobj
16 0 obj
<<
/Names [ (\057\137\137WKANCHOR\1372) [ 0 /XYZ 36 754 0.0 ] (\057\137\137WKANCHOR\1374) [ 0 /XYZ 305 754 0.0 ] (\057\137\137WKANCHOR\1376) [ 0 /XYZ 36 454 0.0 ] ]
>>
```

--- a/pypdf/_doc_common.py
+++ b/pypdf/_doc_common.py
@@ -855,7 +855,7 @@ class PdfDocCommon:
                 # §12.3.3 Document outline, entries in the outline dictionary
                 if not is_null_or_none(lines) and "/First" in lines:
                     node = cast(DictionaryObject, lines["/First"])
-            self._namedDests = self._get_named_destinations()
+            self._named_destinations = self._get_named_destinations()
 
         if node is None:
             return outline
@@ -996,7 +996,7 @@ class PdfDocCommon:
             # TODO : keep named destination instead of replacing it ?
             try:
                 outline_item = self._build_destination(
-                    title, self._namedDests[dest].dest_array
+                    title, self._named_destinations[dest].dest_array
                 )
             except KeyError:
                 # named destination not found in Name Dict

--- a/pypdf/_writer.py
+++ b/pypdf/_writer.py
@@ -2702,7 +2702,10 @@ class PdfWriter(PdfDocCommon):
         reader._namedDests = (
             reader.named_destinations
         )  # need for the outline processing below
-        for dest in reader._namedDests.values():
+
+        arr: Any
+
+        def _process_named_dests(dest: Any) -> None:
             arr = dest.dest_array
             if "/Names" in self._root_object and dest["/Title"] in cast(
                 List[Any],
@@ -2718,7 +2721,10 @@ class PdfWriter(PdfDocCommon):
             elif isinstance(dest["/Page"], int):
                 # the page reference is a page number normally not a PDF Reference
                 # page numbers as int are normally accepted only in external goto
-                p = reader.pages[dest["/Page"]]
+                try:
+                    p = reader.pages[dest["/Page"]]
+                except IndexError:
+                    return
                 assert p.indirect_reference is not None
                 try:
                     arr[NumberObject(0)] = NumberObject(
@@ -2732,6 +2738,9 @@ class PdfWriter(PdfDocCommon):
                     dest["/Page"].indirect_reference.idnum
                 ].indirect_reference
                 self.add_named_destination_array(dest["/Title"], arr)
+
+        for dest in reader._namedDests.values():
+            _process_named_dests(dest)
 
         outline_item_typ: TreeObject
         if outline_item is not None:

--- a/pypdf/_writer.py
+++ b/pypdf/_writer.py
@@ -2699,7 +2699,7 @@ class PdfWriter(PdfDocCommon):
                 position += 1
             srcpages[pg.indirect_reference.idnum].original_page = pg
 
-        reader._namedDests = (
+        reader._named_destinations = (
             reader.named_destinations
         )  # need for the outline processing below
 
@@ -2739,7 +2739,7 @@ class PdfWriter(PdfDocCommon):
                 ].indirect_reference
                 self.add_named_destination_array(dest["/Title"], arr)
 
-        for dest in reader._namedDests.values():
+        for dest in reader._named_destinations.values():
             _process_named_dests(dest)
 
         outline_item_typ: TreeObject

--- a/tests/test_writer.py
+++ b/tests/test_writer.py
@@ -2468,3 +2468,15 @@ def test_increment_writer(caplog):
     assert writer.metadata is None
     b = BytesIO()
     writer.write(b)
+
+
+@pytest.mark.enable_socket()
+def test_append_pdf_with_dest_without_page(caplog):
+    """Tests for #2842"""
+    url = "https://github.com/user-attachments/files/16990834/test.pdf"
+    name = "iss2842.pdf"
+    reader = PdfReader(BytesIO(get_data_from_url(url, name=name)))
+    writer = PdfWriter()
+    writer.append(reader)
+    assert "/__WKANCHOR_8" not in writer.named_destinations
+    assert len(writer.named_destinations) == 3

